### PR TITLE
Support PEP 563 for HfArgumentParser

### LINKED ISSUE #15739: Add compatibility for Postponed Evaluation of Annotations (PEP 563)
Hello,

The code says that it will add compatibility for Postponed Evaluation of Annotations ([PEP 563](https://www.python.org/dev/peps/pep-0563/)) when Python 3.9 is released (which already happened on 2020.10.5). Is there any plan to complete this?

https://github.com/huggingface/transformers/blob/2c2a31ffbcfe03339b1721348781aac4fc05bc5e/src/transformers/hf_argparser.py#L85-L90

### PR DESCRIPTION
**Pull request by function2-llx**
# What does this PR do?

Fixes #15739, support for [PEP 563](https://www.python.org/dev/peps/pep-0563/) is added, making it possible to write `from __future__ import annotations` in the file of dataclasses for `HfArgumentParser`. But this will not be available for Python with a version lower than 3.7, thus adding `from __future__ import annotations` in the test is currently infeasible because `__future__` imports should always occur at the beginning of the file. Instead, I write a test case with string literal annotations.

## Before submitting
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/master/CONTRIBUTING.md#start-contributing-pull-requests), Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link to it if that's the case.
- [x] Did you make sure to update the documentation with your changes?
- [x] Did you write any new necessary tests?

## Who can review?
@LysandreJik

### CODE REVIEW

File: src/transformers/hf_argparser.py [hunk 1]
```diff
@@ -70,93 +70,100 @@ def __init__(self, dataclass_types: Union[DataClassType, Iterable[DataClassType]
...
+            if field.type is bool or (field.default is not None and field.default is not dataclasses.MISSING):
+                # Default value is False if we have no default when of type bool.
+                default = False if field.default is dataclasses.MISSING else field.default
+                # This is the value that will get picked if we don't include --field_name in any way
+                kwargs["default"] = default
+                # This tells argparse we accept 0 or 1 value after --field_name
+                kwargs["nargs"] = "?"
+                # This is the value that will get picked if we do --field_name (without value)
+                kwargs["const"] = True
+        elif isclass(origin_type) and issubclass(origin_type, list):
```
Comment (function2-llx) [inline src/transformers/hf_argparser.py:122]: An alternative implementation without regular expression for #15951

**Comment by function2-llx** [comment]
@LysandreJik Hello, would be great if you could help or ask someone to give any feedback on this, thanks!
Seems #15951 updated the parser as well (just resolved the conflict), I've done the same work, and added a comment at the corresponding changing line.

**Comment by function2-llx** [comment]
Looking forward to any feedback.

**Comment by LysandreJik** [comment]
Thanks for the ping! @sgugger could you give this a quick look? It's similar to the work merged in https://github.com/huggingface/transformers/pull/15951

**Comment by sgugger** [comment]
It looks like you did a rebase on your PR that GitHub did not appreciate (the diff shows 292 files). Could you close this PR and open a fresh one?

File: src/transformers/hf_argparser.py [hunk 2]
```diff
@@ -70,93 +70,100 @@ def __init__(self, dataclass_types: Union[DataClassType, Iterable[DataClassType]
...
+
+        origin_type = getattr(field.type, "__origin__", field.type)
+        if origin_type is Union:
+            if len(field.type.__args__) != 2 or type(None) not in field.type.__args__:
+                raise ValueError("Only `Union[X, NoneType]` (i.e., `Optional[X]`) is allowed for `Union`")
+            if bool not in field.type.__args__:
+                # filter `NoneType` in Union (except for `Union[bool, NoneType]`)
+                field.type = next(
+                    filter(lambda t: not isinstance(None, getattr(t, "__origin__", t)), field.type.__args__)
+                )
```
Comment (sgugger) [inline src/transformers/hf_argparser.py:93]: This line is way to hard to follow. At this stage we know we have only two types, one being NoneType so I think we just want to do
```suggestion
                field.type = field.type.__args__[0] if field.type.__args__[0] != type(None) else field.type.__args__[1]
```

**Review by sgugger — APPROVED** [review]
Thanks a lot for working on this! The diff is indeed more readable now.
I just have one comment to make the code more readable, otherwise LGTM!

**Comment by function2-llx** [comment]
@sgugger Thanks so much for the feedback! I've adopted that. In addition, I've replaced `... != type(None)` with `isinstance` to pass the quality check.

**Comment by sgugger** [comment]
Cool, thanks again for your contribution!
